### PR TITLE
FSPT-725: Build 'preview report' journey for Deliver Grant Funding reports tab

### PR DESCRIPTION
**Pull request by MarcUsher**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-725

## 📝 Description
As we continue to build out the reports tab, Grant Team members will need to be able to preview reports and tasks that have been created in their grant.

This lifts adds a Deliver Grant Funding instance of the Form Runner and hooks it up to the reports tab to preview reports and tasks. The runner functionality is unchanged, except for the 'return to task/report builder' redirects which now redirect to the Reports tab rather than the Developers tab.

It also adds the usual grant team member minimum permissions to these `preview report` routes.

## 📸 Show the thing (screenshots, gifs)
Gif to come

## 🧪 Testing
- Tested as Funding Service admin (preview reports/tasks from the reports tab, all expected test submission functionality and form runner is there)
- Tested as 'Grant team member' (preview reports/tasks etc.)

### Review Readiness
- [X] PR title and description provide sufficient context for reviewers
- [X] Commits are logical, self-contained, and have clear descriptions

### Performance and security
- ~[ ] No N+1 query problems introduced~
- [X] Any new DB queries include appropriate where clauses based on the user's permissions

### Testing
- [X] I have tested this change and it meets the acceptance criteria for the ticket
- [X] I need the reviewer(s) to pull and run this change locally
- [X] New (non-developer) functionality has appropriate unit and integration tests
- ~[ ] End-to-end tests have been updated (if applicable)~ will add e2e tests for the whole Reports tab journey separately
- [X] Edge cases and error conditions are tested

### CODE REVIEW

File: app/common/data/interfaces/collections.py [hunk 1]
```diff
@@ -663,3 +663,19 @@ def delete_form(form: Form) -> None:
...
+        .join(Collection)
+        .where(
+            Collection.id == collection.id,
+            Submission.created_by_id == created_by_user.id,
+        )
+        .all()
+    )
+
+    for submission in submissions:
+        db.session.delete(submission)
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply app/common/data/interfaces/collections.py:680]: Ah yep sure thing!

File: tests/integration/common/data/interfaces/test_collections.py [hunk 2]
```diff
@@ -1194,3 +1205,52 @@ def test_form_reordering(self, db_session, factories):
 
         assert [f.order for f in section.forms] == [0, 1, 2, 3]
         assert section.forms == [forms[0], forms[1], forms[3], forms[4]]
+
+
+class TestDeleteCollectionSubmissions:
```
Comment (samuelhwilliams) [inline tests/integration/common/data/interfaces/test_collections.py:1210]: Nice test for this 👍

File: app/deliver_grant_funding/helpers.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from app.common.data import interfaces
+from app.common.data.interfaces.collections import create_submission, delete_collection_test_submissions_created_by_user
+from app.common.data.types import SubmissionModeEnum
+from app.common.helpers.collections import SubmissionHelper
+
+if TYPE_CHECKING:
+    from app.common.data.models import Collection, Form
+
+
+def start_testing_submission(collection: "Collection", form: Optional["Form"] = None) -> ResponseReturnValue:
```
Comment (samuelhwilliams) [inline app/deliver_grant_funding/helpers.py:15]: 1 small test for this:
- pre-existing test submission that gets deleted/wiped
- redirects go to expected URLs for form-based vs task-based

File: app/deliver_grant_funding/routes/reports.py [hunk 4]
```diff
@@ -111,15 +112,25 @@ def change_report_name(grant_id: UUID, report_id: UUID) -> ResponseReturnValue:
...
 def list_report_tasks(grant_id: UUID, report_id: UUID) -> ResponseReturnValue:
     report = get_collection(report_id, grant_id=grant_id, type_=CollectionType.MONITORING_REPORT, with_full_schema=True)
+    form = GenericSubmitForm()
 
     if report.has_non_default_sections:
         raise RuntimeError(f"Report {report_id} has non-default sections - `add_task` needs updating to handle this.")
 
-    return render_template("deliver_grant_funding/reports/list_report_tasks.html", grant=report.grant, report=report)
+    if form.validate_on_submit() and form.submit.data:
+        return start_testing_submission(collection=report)
```
Comment (samuelhwilliams) [inline app/deliver_grant_funding/routes/reports.py:126]: Probably a test in `TestListReportTasks` for POSTing the form - can just check that it does the redirect to the expected place?

File: app/deliver_grant_funding/routes/runner.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+
+@deliver_grant_funding_blueprint.route(
+    "/grant/<uuid:grant_id>/submissions/<uuid:submission_id>", methods=["GET", "POST"]
+)
+@auto_commit_after_request
+@has_grant_role(RoleEnum.MEMBER)
+def submission_tasklist(grant_id: UUID, submission_id: UUID) -> ResponseReturnValue:
+    source = request.args.get("source")
+    runner = DGFFormRunner.load(submission_id=submission_id, source=FormRunnerState(source) if source else None)
```
Comment (samuelhwilliams) [inline app/deliver_grant_funding/routes/runner.py:21]: These pages will also need to start validating that the user has access to the given submission. We've skipped this for the developer pages, but for the real pages I think we can't ignore this anymore.

For test submissions this means that user A can't access user B's test submission. For live submissions, it will mean that anyone in Org A can't access Org B's submissions.

We'll need to think about the best place to put this - probably in the FormRunner as part of retrieving the submission via `get_submission` - might be that `get_submission` _requires_ passing in a user and that's fed through to the SQL queries to do the checks there.
